src/main/scala/internal/Graph.scala

```diff
@@ -0,0 +1,425 @@
+   */
+  def getOperator(id: NodeId): Operator = operators(id)
+
+  /**
+   * Get {@param num} [[NodeId]]s that don't clash with those of existing nodes.
+   */
+  private def nextNodeIds(num: Int): Seq[NodeId] = {
+    val maxId = (nodes.map(_.id) + 0).max
+    (1 to num).map(i => NodeId(maxId + i))
+  }
+
+  /**
+   * Get {@param num} [[SourceId]]s that don't clash with those of existing sources.
+   */
+  private def nextSourceIds(num: Int): Seq[SourceId] = {
+    val maxId = (sources.map(_.id) + 0).max
+    (1 to num).map(i => SourceId(maxId + i))
+  }
+
+  /**
+   * Get {@param num} [[SinkId]]s that don't clash with those of existing sinks.
+   */
+  private def nextSinkIds(num: Int): Seq[SinkId] = {
+    val maxId = (sinks.map(_.id) + 0).max
+    (1 to num).map(i => SinkId(maxId + i))
+  }
+
+  /**
+   * Get a [[NodeId]] that doesn't clash with any existing node in this graph.
+   */
+  private def nextNodeId(): NodeId = nextNodeIds(1).head
+
+  /**
+   * Get a [[SourceId]] that doesn't clash with any existing source in this graph.
+   */
+  private def nextSourceId(): SourceId = nextSourceIds(1).head
+
+  /**
+   * Get a [[SinkId]] that doesn't clash with any existing sink in this graph.
+   */
+  private def nextSinkId(): SinkId = nextSinkIds(1).head
+
+  /**
+   * Immutably add a node to this graph.
+   *
+   * @param op The operator to be stored at the new node
+   * @param deps The dependencies of the new node
+   * @return A pair containing the new graph and the id assigned to the new node
+   */
+  def addNode(op: Operator, deps: Seq[NodeOrSourceId]): (Graph, NodeId) = {
+    require ({
+      val nodesAndSources: Set[NodeOrSourceId] = nodes ++ sources
+      deps.forall(dep => nodesAndSources.contains(dep))
+    }, "Node must have dependencies on existing ids")
+
+    val id = nextNodeId()
+    val newOperators = operators.updated(id, op)
+    val newDependencies = dependencies.updated(id, deps)
+    (copy(operators = newOperators, dependencies = newDependencies), id)
+  }
+
+  /**
+   * Immutably add a new sink to this graph.
+   *
+   * @param dep The dependency of the new sink
+   * @return A pair containing the new graph and the id assigned to the new sink
+   */
+  def addSink(dep: NodeOrSourceId): (Graph, SinkId) = {
+    require ({
+      val nodesAndSources: Set[NodeOrSourceId] = nodes ++ sources
+      nodesAndSources.contains(dep)
+    }, "Sink must have dependencies on an existing id")
+
+    val id = nextSinkId()
+    val newSinkDependencies = sinkDependencies.updated(id, dep)
+    (copy(sinkDependencies = newSinkDependencies), id)
+  }
+
+  /**
+   * Immutably add a new source to this graph.
+   *
+   * @return A pair containing the new graph and the id assigned to the new source
+   */
+  def addSource(): (Graph, SourceId) = {
+    val id = nextSourceId()
+    val newSources = sources + id
+    (copy(sources = newSources), id)
+  }
+
+  /**
+   * Immutably update the dependencies of a node in this graph.
```

Comment:
"Immutably update" is kind of an oxymoron - maybe there's a better way to phrase this?